rest_tools/server/handler.py

```diff
@@ -351,3 +357,114 @@ async def wrapper(self, *args, **kwargs):
+            'redirect_uri': redirect_uri,
+            'code': code,
+            'client_id': self.oauth_client_id,
+            'client_secret': self.oauth_client_secret,
+            'grant_type': 'authorization_code',
+        })
+        response = await http.fetch(
+            self._OAUTH_ACCESS_TOKEN_URL,
+            method='POST',
+            headers={'Content-Type': 'application/x-www-form-urlencoded'},
+            body=body,
+        )
+        ret = tornado.escape.json_decode(response.body)
+        if not ret.get('id_token', ''):
+            response = await http.fetch(
+                self._OAUTH_USERINFO_URL,
+                method='GET',
+                headers={'Authorization': f'Bearer {ret["access_token"]}'},
+            )
+            ret['id_token'] = tornado.escape.json_decode(response.body)
+
+        self.auth.validate(ret['access_token'])
+        return ret
+
+    def _decode_state(self, state: Union[bytes, str]) -> str:
+        data = tornado.escape.json_decode(base64.b64decode(state))
+        _, token, _ = self._decode_xsrf_token(data.pop('xsrf'))
+        _, expected_token, _ = self._get_raw_xsrf_token()
+        if not token:
+            raise tornado.web.HTTPError(403, reason="'state' argument has invalid format")
+        if not hmac.compare_digest(tornado.escape.utf8(token), tornado.escape.utf8(expected_token)):
+            raise tornado.web.HTTPError(403, reason="XSRF cookie does not match state argument")
+        return data
+
+    def _encode_state(self, data: Dict[str, Any]) -> bytes:
+        data2 = data.copy()  # make a copy to not add xsrf to source dict
+        data2['xsrf'] = self.xsrf_token.decode('utf-8')
+        return base64.b64encode(tornado.escape.json_encode(data2).encode('utf-8'))
+
+    @catch_error
+    async def get(self):
+        if self.get_argument('error', False):
+            err = self.get_argument('error_description', None)
+            if not err:
+                err = 'unknown oauth2 error'
+            raise tornado.web.HTTPError(400, reason=err)
+        elif self.get_argument('code', False):
+            data = self._decode_state(self.get_argument('state'))
+            user = await self.get_authenticated_user(
+                redirect_uri=self.get_login_url(),
+                code=self.get_argument('code'),
+            )
+            # Save the user with e.g. set_secure_cookie
+            self.set_secure_cookie('access_token', user['access_token'],
+                                   expires_days=1.0*user['expires_in']/3600/24)
+            if 'refresh_token' in user:
+                self.set_secure_cookie('refresh_token', user['refresh_token'],
+                                       expires_days=1.0*user.get('refresh_expires_in', 86400)/3600/24)
+            self.set_secure_cookie('identity', tornado.escape.json_encode(user['id_token']),
+                                   expires_days=1.0*user.get('refresh_expires_in', 86400)/3600/24)
```

Comment:
Using `1.0*int_var` as a float cast scares me a bit. this could be lost in future refactoring. Is there any reason why this shouldn't be `float(int_var)`?

Reply:
No, that's a good point.  I actually copied this part from some older code I had, which was "dirty."  Will fix.